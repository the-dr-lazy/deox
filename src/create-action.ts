export type Action<
  TType extends string,
  TPayload = undefined,
  TMeta = undefined
> = TPayload extends undefined
  ? (TMeta extends undefined ? { type: TType } : { type: TType; meta: TMeta })
  : (TPayload extends Error
      ? (TMeta extends undefined
          ? { type: TType; payload: TPayload; error: true }
          : { type: TType; payload: TPayload; meta: TMeta; error: true })
      : (TMeta extends undefined
          ? { type: TType; payload: TPayload }
          : { type: TType; payload: TPayload; meta: TMeta }))
export type AnyAction = Action<string>

/**
 * Minimal (type-only) action factory
 * @example
 * const clearTodos = action('[Todo] truncate');
 */
export function createAction<TType extends string>(type: TType): Action<TType>
/**
 * Action with error factory
 * @example
 * const fetchTodosRejected = (payload: Error) => action('[Todo] fetch rejected', payload);
 */
export function createAction<TType extends string, TPayload extends Error>(
  type: TType,
  payload: TPayload
): Action<TType, TPayload>
/**
 * Action with non-error payload factory
 * @example
 * const addTodo = ({ name, completed = false }: Todo) => action('[Todo] add', { name, completed });
 */
export function createAction<TType extends string, TPayload>(
  type: TType,
  payload: TPayload
): Action<TType, TPayload>
/**
 * Action with error and meta factory
 * @example
 * const fetchTodosRejected = (payload: Error, meta?: Meta) => action('[Todo] fetch rejected', payload, meta);
 */
export function createAction<
  TType extends string,
  TPayload extends Error,
  TMeta
>(type: TType, payload: TPayload, meta: TMeta): Action<TType, TPayload, TMeta>
/**
 * Action with payload and meta factory
 * @example
 * const addTodo = ({ name, completed = false }: Todo, meta?: Meta) => action('[Todo] add', { name, completed }, meta);
 */
export function createAction<TType extends string, TPayload, TMeta>(
  type: TType,
  payload: TPayload,
  meta: TMeta
): Action<TType, TPayload, TMeta>

/**
 * Flux standard action factory
 * @example
 * const clearTodos = action('[Todo] truncate');
 * @example
 * const fetchTodosRejected = (payload: Error) => action('[Todo] fetch rejected', payload);
 * @example
 * const addTodo = ({ name, completed = false }: Todo) => action('[Todo] add', { name, completed });
 * @example
 * const fetchTodosRejected = (payload: Error, meta?: Meta) => action('[Todo] fetch rejected', payload, meta);
 * @example
 * const addTodo = ({ name, completed = false }: Todo, meta?: Meta) => action('[Todo] add', { name, completed }, meta);
 */
export function createAction<TType extends string, TPayload, TMeta>(
  type: TType,
  payload?: TPayload,
  meta?: TMeta
) {
  return {
    type,
    ...(payload !== undefined ? { payload } : {}),
    ...(meta !== undefined ? { meta } : {}),
    ...(payload instanceof Error ? { error: true } : {}),
  }
}
