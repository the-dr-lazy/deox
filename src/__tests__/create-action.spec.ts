import { createAction } from '../create-action'

describe('action', () => {
  it.each`
    name                            | type                       | payload                      | meta
    ${'type-only'}                  | ${'[Todo] truncate'}       | ${undefined}                 | ${undefined}
    ${'error payload'}              | ${'[Todo] fetch rejected'} | ${new Error()}               | ${undefined}
    ${'error and meta'}             | ${'[Todo] fetch rejected'} | ${new Error()}               | ${{ status: 403 }}
    ${'non-error payload'}          | ${'[Todo] add'}            | ${{ name: `Read Robbin's` }} | ${undefined}
    ${'non-error payload and meta'} | ${'[Todo] add'}            | ${{ name: `Read Robbin's` }} | ${'Book is on bookshelf'}
    ${'`false` payload'}            | ${'[Todo] show'}           | ${false}                     | ${undefined}
    ${'`null` payload'}             | ${'[Todo] show'}           | ${null}                      | ${undefined}
    ${'`0` payload'}                | ${'[Todo] show'}           | ${0}                         | ${undefined}
    ${'`NaN` payload'}              | ${'[Todo] show'}           | ${NaN}                       | ${undefined}
    ${"`''` payload"}               | ${'[Todo] show'}           | ${''}                        | ${undefined}
    ${'`false` meta'}               | ${'[Todo] show'}           | ${undefined}                 | ${false}
    ${'`null` meta'}                | ${'[Todo] show'}           | ${undefined}                 | ${null}
    ${'`0` meta'}                   | ${'[Todo] show'}           | ${undefined}                 | ${0}
    ${'`NaN` meta'}                 | ${'[Todo] show'}           | ${undefined}                 | ${NaN}
    ${"`''` meta"}                  | ${'[Todo] show'}           | ${undefined}                 | ${''}
  `('should create action with $name', ({ type, payload, meta }) => {
    expect(createAction(type, payload, meta)).toMatchSnapshot()
  })
})
