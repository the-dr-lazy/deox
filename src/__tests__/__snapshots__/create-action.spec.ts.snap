// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action should create action with \`''\` meta 1`] = `
Object {
  "meta": "",
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`''\` payload 1`] = `
Object {
  "payload": "",
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`0\` meta 1`] = `
Object {
  "meta": 0,
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`0\` payload 1`] = `
Object {
  "payload": 0,
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`NaN\` meta 1`] = `
Object {
  "meta": NaN,
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`NaN\` payload 1`] = `
Object {
  "payload": NaN,
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`false\` meta 1`] = `
Object {
  "meta": false,
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`false\` payload 1`] = `
Object {
  "payload": false,
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`null\` meta 1`] = `
Object {
  "meta": null,
  "type": "[Todo] show",
}
`;

exports[`action should create action with \`null\` payload 1`] = `
Object {
  "payload": null,
  "type": "[Todo] show",
}
`;

exports[`action should create action with error and meta 1`] = `
Object {
  "error": true,
  "meta": Object {
    "status": 403,
  },
  "payload": [Error],
  "type": "[Todo] fetch rejected",
}
`;

exports[`action should create action with error payload 1`] = `
Object {
  "error": true,
  "payload": [Error],
  "type": "[Todo] fetch rejected",
}
`;

exports[`action should create action with non-error payload 1`] = `
Object {
  "payload": Object {
    "name": "Read Robbin's",
  },
  "type": "[Todo] add",
}
`;

exports[`action should create action with non-error payload and meta 1`] = `
Object {
  "meta": "Book is on bookshelf",
  "payload": Object {
    "name": "Read Robbin's",
  },
  "type": "[Todo] add",
}
`;

exports[`action should create action with type-only 1`] = `
Object {
  "type": "[Todo] truncate",
}
`;
